File: sdk/typescript/tests/agui.test.ts
import { describe, it, expect } from 'vitest';
import { agui } from '../src/index.js';
import type { ToolCallTrace } from '../src/ai/ToolCalling.js';

describe('agui — buffered helpers', () => {
  it('toolCall builds the canonical entry', () => {
    expect(agui.toolCall('showFlightCard', { from: 'SFO', to: 'JFK' })).toEqual({
      name: 'showFlightCard',
      arguments: { from: 'SFO', to: 'JFK' },
    });
  });

  it('toolCall handles empty arguments and explicit id', () => {
    expect(agui.toolCall('ping', undefined, { id: 'tc-1' })).toEqual({
      id: 'tc-1',
      name: 'ping',
      arguments: {},
    });
  });

  it('toolCall surfaces a result when provided', () => {
    expect(agui.toolCall('lookup', { q: 'x' }, { result: { ok: true } })).toEqual({
      name: 'lookup',
      arguments: { q: 'x' },
      result: { ok: true },
    });
  });

  it('toolCall hasResult forces a null result through', () => {
    expect(agui.toolCall('noop', undefined, { hasResult: true })).toEqual({
      name: 'noop',
      arguments: {},
      result: undefined,
    });
  });

  it('toolCallsFromTrace returns [] for empty / null traces', () => {
    expect(agui.toolCallsFromTrace(null)).toEqual([]);
    expect(agui.toolCallsFromTrace(undefined)).toEqual([]);
    const empty: ToolCallTrace = { calls: [], totalTurns: 0, totalToolCalls: 0 };
    expect(agui.toolCallsFromTrace(empty)).toEqual([]);
  });

  it('toolCallsFromTrace converts records, surfaces result and error', () => {
    const trace: ToolCallTrace = {
      totalTurns: 1,
      totalToolCalls: 2,
      calls: [
        { toolName: 'a', arguments: { x: 1 }, result: { ok: true }, latencyMs: 5, turn: 0 },
        { toolName: 'b', arguments: {}, error: 'boom', latencyMs: 5, turn: 0 },
      ],
    };
    expect(agui.toolCallsFromTrace(trace)).toEqual([
      { id: 'tc-trace-0', name: 'a', arguments: { x: 1 }, result: { ok: true } },
      { id: 'tc-trace-1', name: 'b', arguments: {}, result: { error: 'boom' } },
    ]);
  });

  it('stateDeltaReplace emits a JSON Patch op', () => {
    expect(agui.stateDeltaReplace('/counter', 2)).toEqual({
      op: 'replace',
      path: '/counter',
      value: 2,
    });
  });

  it('stateDeltaReplace rejects paths missing a leading slash', () => {
    expect(() => agui.stateDeltaReplace('counter', 1)).toThrow(/RFC 6902/);
  });

  it('stateDeltaFromDiff emits a minimal shallow patch', () => {
    const before = { a: 1, b: 2, c: 3 };
    const after = { a: 1, b: 99, d: 4 };
    expect(agui.stateDeltaFromDiff(before, after)).toEqual([
      { op: 'replace', path: '/b', value: 99 },
      { op: 'remove', path: '/c' },
      { op: 'add', path: '/d', value: 4 },
    ]);
  });

  it('reasoningSegment + reasoning() build the segment list', () => {
    const seg = agui.reasoningSegment('thinking', { id: 'r1' });
    expect(seg).toEqual({ content: 'thinking', id: 'r1' });
    expect(agui.reasoning('a', '', seg, 'b')).toEqual([
      'a',
      { content: 'thinking', id: 'r1' },
      'b',
    ]);
  });

  it('reasoning() rejects garbage segments', () => {
    expect(() => agui.reasoning(42 as unknown as string)).toThrow(/segments must be string/);
  });
});

describe('agui — streaming chunk builders', () => {
  it('text/reasoning/reasoning_end', () => {
    expect(agui.textChunk('hi')).toEqual({ type: 'text', delta: 'hi' });
    expect(agui.reasoningChunk('thinking')).toEqual({ type: 'reasoning', delta: 'thinking' });
    expect(agui.reasoningEndChunk()).toEqual({ type: 'reasoning_end' });
  });

  it('toolCallStart with and without args/parent', () => {
    expect(agui.toolCallStartChunk('tc1', 'foo')).toEqual({
      type: 'tool_call_start',
      id: 'tc1',
      name: 'foo',
    });
    expect(
      agui.toolCallStartChunk('tc2', 'bar', { arguments: { x: 1 }, parentMessageId: 'm1' }),
    ).toEqual({
      type: 'tool_call_start',
      id: 'tc2',
      name: 'bar',
      arguments: { x: 1 },
      parentMessageId: 'm1',
    });
  });

  it('toolCallArgs / toolCallEnd / toolCallResult', () => {
    expect(agui.toolCallArgsChunk('tc1', '{"x":')).toEqual({
      type: 'tool_call_args',
      id: 'tc1',
      delta: '{"x":',
    });
    expect(agui.toolCallEndChunk('tc1')).toEqual({ type: 'tool_call_end', id: 'tc1' });
    expect(agui.toolCallResultChunk('tc1', 'done')).toEqual({
      type: 'tool_call_result',
      id: 'tc1',
      content: 'done',
      role: 'tool',
    });
    expect(agui.toolCallResultChunk('tc1', 'done', { role: 'system' })).toMatchObject({
      role: 'system',
    });
  });

  it('state / state_delta', () => {
    expect(agui.stateChunk({ k: 1 })).toEqual({ type: 'state', snapshot: { k: 1 } });
    const ops = [agui.stateDeltaReplace('/k', 2)];
    expect(agui.stateDeltaChunk(ops)).toEqual({ type: 'state_delta', ops });
  });

  it('step_started / step_finished', () => {
    expect(agui.stepStartedChunk('plan')).toEqual({ type: 'step_started', name: 'plan' });
    expect(agui.stepFinishedChunk('plan')).toEqual({ type: 'step_finished', name: 'plan' });
  });

  it('raw / custom / final / error chunk shapes', () => {
    expect(agui.rawChunk({ k: 1 })).toEqual({ type: 'raw', event: { k: 1 } });
    expect(agui.rawChunk({ k: 1 }, { source: 'harness' })).toEqual({
      type: 'raw',
      event: { k: 1 },
      source: 'harness',
    });
    expect(agui.customChunk('progress', 0.5)).toEqual({
      type: 'custom',
      name: 'progress',
      value: 0.5,
    });
    expect(agui.customChunk('ping')).toEqual({ type: 'custom', name: 'ping' });
    expect(agui.finalChunk({ result: 'done' })).toEqual({
      type: 'final',
      data: { result: 'done' },
    });
    expect(agui.errorChunk('boom')).toEqual({ type: 'error', message: 'boom' });
    expect(agui.errorChunk('boom', { code: 'E_BOOM' })).toEqual({
      type: 'error',
      message: 'boom',
      code: 'E_BOOM',
    });
  });
});

describe('agui — serializeStream', () => {
  it('emits one NDJSON line per chunk', async () => {
    async function* chunks() {
      yield agui.textChunk('a');
      yield agui.textChunk('b');
      yield agui.toolCallEndChunk('tc1');
    }
    const decoder = new TextDecoder();
    const lines: string[] = [];
    for await (const buf of agui.serializeStream(chunks())) {
      lines.push(decoder.decode(buf));
    }
    expect(lines).toEqual([
      JSON.stringify({ type: 'text', delta: 'a' }) + '\n',
      JSON.stringify({ type: 'text', delta: 'b' }) + '\n',
      JSON.stringify({ type: 'tool_call_end', id: 'tc1' }) + '\n',
    ]);
  });

  it('auto-wraps bare strings as text chunks', async () => {
    async function* gen() {
      yield 'hello';
      yield agui.textChunk(' world');
    }
    const decoder = new TextDecoder();
    let combined = '';
    for await (const buf of agui.serializeStream(gen())) combined += decoder.decode(buf);
    expect(combined.trim().split('\n').map((l) => JSON.parse(l))).toEqual([
      { type: 'text', delta: 'hello' },
      { type: 'text', delta: ' world' },
    ]);
  });

  it('rejects non-string non-object values', async () => {
    async function* gen() {
      yield 42 as unknown as string;
    }
    await expect(async () => {
      for await (const _ of agui.serializeStream(gen())) {
        /* drain */
      }
    }).rejects.toThrow(/non-string\/non-object/);
  });

  it('accepts a synchronous iterable too', async () => {
    const chunks = [agui.textChunk('x'), agui.textChunk('y')];
    const decoder = new TextDecoder();
    let n = 0;
    for await (const buf of agui.serializeStream(chunks)) {
      const obj = JSON.parse(decoder.decode(buf));
      expect(obj.type).toBe('text');
      n++;
    }
    expect(n).toBe(2);
  });
});

describe('agui — relayHarnessStream', () => {
  async function* fromArray(items: unknown[]) {
    for (const x of items) yield x;
  }

  it('translates assistant text blocks into text chunks', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(
      fromArray([
        {
          type: 'assistant',
          message: {
            content: [
              { type: 'text', text: 'hello ' },
              { type: 'text', text: 'world' },
            ],
          },
        },
      ]),
    )) {
      chunks.push(ch);
    }
    expect(chunks).toEqual([
      { type: 'text', delta: 'hello ' },
      { type: 'text', delta: 'world' },
    ]);
  });

  it('translates assistant thinking blocks into reasoning chunks', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(
      fromArray([
        {
          type: 'assistant',
          message: { content: [{ type: 'thinking', thinking: 'hmm' }] },
        },
      ]),
    )) {
      chunks.push(ch);
    }
    expect(chunks).toEqual([{ type: 'reasoning', delta: 'hmm' }]);
  });

  it('translates tool_use blocks into start+end pairs', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(
      fromArray([
        {
          type: 'assistant',
          message: {
            content: [{ type: 'tool_use', id: 'tc1', name: 'lookup', input: { q: 'x' } }],
          },
        },
      ]),
    )) {
      chunks.push(ch);
    }
    expect(chunks).toEqual([
      { type: 'tool_call_start', id: 'tc1', name: 'lookup', arguments: { q: 'x' } },
      { type: 'tool_call_end', id: 'tc1' },
    ]);
  });

  it('translates tool_result string content', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(
      fromArray([
        {
          type: 'user',
          message: {
            content: [{ type: 'tool_result', tool_use_id: 'tc1', content: 'ok' }],
          },
        },
      ]),
    )) {
      chunks.push(ch);
    }
    expect(chunks).toEqual([
      { type: 'tool_call_result', id: 'tc1', content: 'ok', role: 'tool' },
    ]);
  });

  it('translates tool_result list content by stitching text blocks', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(
      fromArray([
        {
          type: 'user',
          message: {
            content: [
              {
                type: 'tool_result',
                tool_use_id: 'tc1',
                content: [
                  { type: 'text', text: 'a' },
                  { type: 'text', text: 'b' },
                ],
              },
            ],
          },
        },
      ]),
    )) {
      chunks.push(ch);
    }
    expect(chunks).toEqual([
      { type: 'tool_call_result', id: 'tc1', content: 'ab', role: 'tool' },
    ]);
  });

  it('skips terminal result envelope and surfaces system as raw', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(
      fromArray([
        { type: 'system', subtype: 'init' },
        { type: 'result', subtype: 'success', result: 'done' },
      ]),
    )) {
      chunks.push(ch);
    }
    expect(chunks).toEqual([
      { type: 'raw', event: { type: 'system', subtype: 'init' }, source: 'harness' },
    ]);
  });

  it('preserves unknown blocks and unknown top-level messages as raw', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(
      fromArray([
        { type: 'assistant', message: { content: [{ type: 'mystery', payload: 1 }] } },
        { type: 'no-such-thing' },
      ]),
    )) {
      chunks.push(ch);
    }
    expect(chunks[0]).toMatchObject({ type: 'raw', source: 'harness' });
    expect((chunks[0] as { event: { type: string } }).event.type).toBe('mystery');
    expect(chunks[1]).toMatchObject({ type: 'raw', source: 'harness' });
  });

  it('handles bare content and string content shapes', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(
      fromArray([
        { type: 'assistant', content: 'inline-string' },
        { type: 'assistant', content: [{ type: 'text', text: 'inline-list' }] },
      ]),
    )) {
      chunks.push(ch);
    }
    expect(chunks).toEqual([
      { type: 'text', delta: 'inline-string' },
      { type: 'text', delta: 'inline-list' },
    ]);
  });

  it('wraps non-object iterates as raw', async () => {
    const chunks = [];
    for await (const ch of agui.relayHarnessStream(fromArray(['scalar', 7, null]))) {
      chunks.push(ch);
    }
    expect(chunks.every((c) => (c as { type: string }).type === 'raw')).toBe(true);
    expect(chunks).toHaveLength(3);
  });
});

describe('agui — STREAMING_CONTENT_TYPE', () => {
  it('matches the wire constant from the Python and Go SDKs', () => {
    expect(agui.STREAMING_CONTENT_TYPE).toBe('application/x-ndjson');
  });
});

File: sdk/go/agent/agui/agui_test.go
package agui

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"strings"
	"testing"
	"time"

	"github.com/Agent-Field/agentfield/sdk/go/ai"

	"github.com/stretchr/testify/require"
)

func TestToolCall_MinimalAndFull(t *testing.T) {
	require.Nil(t, ToolCall("", "", nil, nil), "empty name returns nil so caller surfaces the bug")

	minimal := ToolCall("", "showFlightCard", nil, nil)
	require.Equal(t, "showFlightCard", minimal["name"])
	require.NotContains(t, minimal, "id", "id only present when caller supplies one")
	require.Equal(t, map[string]any{}, minimal["arguments"])
	require.NotContains(t, minimal, "result")

	full := ToolCall("tc-1", "x", map[string]any{"a": 1}, map[string]any{"ok": true})
	require.Equal(t, "tc-1", full["id"])
	require.Equal(t, map[string]any{"a": 1}, full["arguments"])
	require.Equal(t, map[string]any{"ok": true}, full["result"])
}

func TestToolCallsFromTrace(t *testing.T) {
	require.Empty(t, ToolCallsFromTrace(nil))
	require.Empty(t, ToolCallsFromTrace(&ai.ToolCallTrace{}))

	trace := &ai.ToolCallTrace{
		Calls: []ai.ToolCallRecord{
			{ToolName: "getWeather", Arguments: map[string]any{"city": "SF"}, Result: map[string]any{"temp": 62.0}},
			{ToolName: "lookup", Arguments: map[string]any{"q": "x"}, Error: "timeout"},
			{ToolName: "noargs"},
		},
	}
	out := ToolCallsFromTrace(trace)
	require.Len(t, out, 3)

	require.Equal(t, "tc-trace-0", out[0]["id"])
	require.Equal(t, "getWeather", out[0]["name"])
	require.Equal(t, map[string]any{"temp": 62.0}, out[0]["result"])

	require.Equal(t, "tc-trace-1", out[1]["id"])
	require.Equal(t, map[string]any{"error": "timeout"}, out[1]["result"], "errors surface as {error:...}")

	require.Equal(t, "tc-trace-2", out[2]["id"])
	require.Equal(t, map[string]any{}, out[2]["arguments"], "nil arguments default to empty map")
	require.NotContains(t, out[2], "result", "no result and no error means omit the field")
}

func TestStateDeltaReplace(t *testing.T) {
	op, err := StateDeltaReplace("/counter", 2)
	require.NoError(t, err)
	require.Equal(t, map[string]any{"op": "replace", "path": "/counter", "value": 2}, op)

	_, err = StateDeltaReplace("counter", 2)
	require.Error(t, err, "path without leading slash is invalid")
	_, err = StateDeltaReplace("", 2)
	require.Error(t, err, "empty path is invalid")
}

func TestStreamingChunkBuilders(t *testing.T) {
	require.Equal(t, map[string]any{"type": "text", "delta": "hi"}, TextChunk("hi"))
	require.Equal(t, map[string]any{"type": "reasoning", "delta": "think"}, ReasoningChunk("think"))
	require.Equal(t, map[string]any{"type": "reasoning_end"}, ReasoningEndChunk())

	tcStart := ToolCallStartChunk("tc1", "x", map[string]any{"a": 1}, "msg-1")
	require.Equal(t, "tool_call_start", tcStart["type"])
	require.Equal(t, "tc1", tcStart["id"])
	require.Equal(t, "x", tcStart["name"])
	require.Equal(t, map[string]any{"a": 1}, tcStart["arguments"])
	require.Equal(t, "msg-1", tcStart["parentMessageId"])

	tcStartNoExtras := ToolCallStartChunk("tc2", "x", nil, "")
	require.NotContains(t, tcStartNoExtras, "arguments")
	require.NotContains(t, tcStartNoExtras, "parentMessageId")

	require.Equal(t, map[string]any{"type": "tool_call_args", "id": "tc1", "delta": "{\"x"}, ToolCallArgsChunk("tc1", "{\"x"))
	require.Equal(t, map[string]any{"type": "tool_call_end", "id": "tc1"}, ToolCallEndChunk("tc1"))

	res := ToolCallResultChunk("tc1", "ok", "")
	require.Equal(t, "tool", res["role"], "default role is 'tool'")
	require.Equal(t, "ok", res["content"])

	require.Equal(t, map[string]any{"type": "state", "snapshot": map[string]any{"a": 1}}, StateChunk(map[string]any{"a": 1}))
	require.Equal(t, "state_delta", StateDeltaChunk([]any{map[string]any{"op": "replace"}})["type"])

	require.Equal(t, "step_started", StepStartedChunk("plan")["type"])
	require.Equal(t, "step_finished", StepFinishedChunk("plan")["type"])

	raw := RawChunk(map[string]any{"x": 1}, "harness")
	require.Equal(t, "raw", raw["type"])
	require.Equal(t, "harness", raw["source"])

	rawNoSrc := RawChunk(map[string]any{"x": 1}, "")
	require.NotContains(t, rawNoSrc, "source")

	custom := CustomChunk("ack", map[string]any{"ok": true})
	require.Equal(t, "custom", custom["type"])
	require.Equal(t, "ack", custom["name"])

	customNil := CustomChunk("ack", nil)
	require.NotContains(t, customNil, "value")

	final := FinalChunk(map[string]any{"toolCalls": []any{}})
	require.Equal(t, "final", final["type"])

	errCh := ErrorChunk("boom", "E1")
	require.Equal(t, "error", errCh["type"])
	require.Equal(t, "boom", errCh["message"])
	require.Equal(t, "E1", errCh["code"])

	errChNoCode := ErrorChunk("boom", "")
	require.NotContains(t, errChNoCode, "code")
}

func TestSerializeStream(t *testing.T) {
	ch := make(chan map[string]any, 4)
	ch <- TextChunk("hello ")
	ch <- TextChunk("world")
	ch <- StateChunk(map[string]any{"counter": 1})
	close(ch)

	var buf bytes.Buffer
	require.NoError(t, SerializeStream(context.Background(), &buf, ch))
	lines := strings.Split(strings.TrimRight(buf.String(), "\n"), "\n")
	require.Len(t, lines, 3)

	var first map[string]any
	require.NoError(t, json.Unmarshal([]byte(lines[0]), &first))
	require.Equal(t, "text", first["type"])
	require.Equal(t, "hello ", first["delta"])

	var third map[string]any
	require.NoError(t, json.Unmarshal([]byte(lines[2]), &third))
	require.Equal(t, "state", third["type"])
}

func TestSerializeStream_RespectsContext(t *testing.T) {
	ch := make(chan map[string]any) // never closed; never sends
	ctx, cancel := context.WithCancel(context.Background())
	done := make(chan error, 1)
	go func() { done <- SerializeStream(ctx, io.Discard, ch) }()
	cancel()
	select {
	case err := <-done:
		require.ErrorIs(t, err, context.Canceled)
	case <-time.After(time.Second):
		t.Fatal("SerializeStream did not honor context cancellation")
	}
}
